### PR TITLE
fix: exclude disconnected and input nodes from model outputs in model_generation.py

### LINKED ISSUE #199: [Bug] model_generation.py crashes with KeyError for disconnected nodes and misidentifies input nodes as outputs
## Describe the bug
`model_generation()` in `app/services/model_generation.py` has two bugs:

**1. KeyError for disconnected nodes**
```python
output_ids = [n["id"] for n in model_params["nodes"] if n["id"] not in source_to_targets]
outputs = [keras_tensors[oid] for oid in output_ids]  # KeyError if node was never visited
```
If a node has no outgoing edges but was never visited during BFS (e.g. disconnected from the graph), it appears in `output_ids` but has no entry in `keras_tensors`, causing a `KeyError` crash.

**2. Input nodes misidentified as outputs**
`custominput` nodes have no outgoing edges in an incomplete graph, so they satisfy `n["id"] not in source_to_targets` and get included in `outputs`. This causes Keras to build a degenerate model where inputs equal outputs.

## Steps to Reproduce
1. Build a canvas with a disconnected node (not connected to the main graph)
2. Submit for validation
3. Observe `KeyError` crash in `model_generation()`

## Expected behavior
- Skip nodes not present in `keras_tensors` when collecting outputs
- Exclude `custominput` nodes from output identification

## Suggested Fix
```python
output_ids = [
    n["id"] for n in model_params["nodes"]
    if n["id"] not in source_to_targets
    and n["type"] != "custominput"
    and n["id"] in keras_tensors
]
```

### PR DESCRIPTION
**Pull request by VishakBaddur**
## Summary
Fixes #199

## Changes
`output_ids` previously identified output nodes as any node with no outgoing edges. This had two problems:

1. **KeyError crash**: disconnected nodes (never visited during BFS) had no entry in `keras_tensors`, causing a crash when collecting outputs
2. **Input nodes misidentified as outputs**: `custominput` nodes with no outgoing edges in an incomplete graph were incorrectly included in outputs, causing Keras to build a degenerate model

Fixed by adding two guards to the output_ids list comprehension:
- `n["type"] != "custominput"`: excludes input nodes
- `n["id"] in keras_tensors`: excludes disconnected unvisited nodes

### CODE REVIEW

**Review by ivantha — COMMENTED** [review]
## ◎ Argus Review

This PR correctly identifies and fixes two real bugs in the output node detection logic: a `KeyError` crash from disconnected nodes and the misidentification of isolated `custominput` nodes as outputs. The fix is minimal and targeted. However, the same `KeyError` vulnerability remains unaddressed in the `inputs` list comprehension on the line directly above, and the fix introduces no tests. There are also several edge cases (e.g., an empty `outputs` list after filtering) that could still cause downstream failures.

File: tensormap-backend/app/services/model_generation.py [hunk 1]
```diff
@@ -66,7 +66,11 @@ def model_generation(model_params: dict) -> dict:
...
     inputs = [keras_tensors[n["id"]] for n in model_params["nodes"] if n["type"] == "custominput"]
-    output_ids = [n["id"] for n in model_params["nodes"] if n["id"] not in source_to_targets]
+    output_ids = [
+        n["id"]
+        for n in model_params["nodes"]
+        if n["id"] not in source_to_targets and n["type"] != "custominput" and n["id"] in keras_tensors
+    ]
     outputs = [keras_tensors[oid] for oid in output_ids]
 
     model = tf.keras.Model(inputs=inputs, outputs=outputs)
```
Comment (ivantha) [inline tensormap-backend/app/services/model_generation.py]: **[HIGH]** Empty `outputs` list after filtering will cause a downstream crash

After the new guards are applied, `output_ids` may legitimately be an empty list in edge cases, such as:
- A graph where all non-input nodes still have outgoing edges (a cycle or incomplete graph)
- A graph where every terminal node is disconnected (not in `keras_tensors`)

This results in `outputs = []`, and then:
```python
model = tf.keras.Model(inputs=inputs, outputs=outputs)
```
will raise a `ValueError` from Keras: *"argument `outputs` must not be empty"*. Similarly, `inputs` could be empty if no `custominput` nodes exist or pass the guard. Both produce an unhandled exception with no meaningful error message surfaced to the user.

**Suggestion:**
```
Add explicit validation before model construction:

```python
if not inputs:
    raise ValueError("Model has no valid input nodes. Ensure at least one 'custominput' node is connected to the graph.")
if not output_ids:
    raise ValueError("Model has no valid output nodes. Ensure at least one terminal (non-input) node is connected to the graph.")
```

This converts silent crashes into actionable error messages that can be surfaced to the frontend.
```

File: tensormap-backend/app/services/model_generation.py [hunk 2]
```diff
@@ -66,7 +66,11 @@ def model_generation(model_params: dict) -> dict:
 
     # Identify input and output tensors
     inputs = [keras_tensors[n["id"]] for n in model_params["nodes"] if n["type"] == "custominput"]
-    output_ids = [n["id"] for n in model_params["nodes"] if n["id"] not in source_to_targets]
+    output_ids = [
+        n["id"]
+        for n in model_params["nodes"]
+        if n["id"] not in source_to_targets and n["type"] != "custominput" and n["id"] in keras_tensors
+    ]
```
Comment (ivantha) [inline tensormap-backend/app/services/model_generation.py]: **[MEDIUM]** No tests added for the new filtering behavior

The PR fixes two distinct crash scenarios but includes no test cases to verify the fix or prevent regressions:

1. A graph with a disconnected node (not in `keras_tensors`) that previously caused a `KeyError`.
2. A graph with an isolated `custominput` node (no outgoing edges) that was previously included in `outputs`.

Without tests, a future refactor of the BFS traversal or the `output_ids` logic could silently reintroduce these bugs.

**Suggestion:**
```
Add unit tests covering at minimum:

```python
def test_output_ids_excludes_custominput_nodes():
    # Graph: input -> dense (no outgoing edges on input node)
    model_params = {"nodes": [...], "edges": [...]}
    result = model_generation(model_params)
    # assert input node not in outputs

def test_output_ids_excludes_disconnected_nodes():
    # Graph with a node present in `nodes` but not reachable from any input
    model_params = {"nodes": [...], "edges": [...]}
    result = model_generation(model_params)  # should not raise KeyError

def test_model_generation_disconnected_input_node():
    # custominput node with no edges — should not crash (covers the remaining bug)
    ...
```
```

**Review by ivantha — COMMENTED** [review]
## ◎ Argus Review

This PR claims to fix two bugs (KeyError crash on disconnected output nodes, and `custominput` nodes being misidentified as outputs) but the actual diff only modifies the `inputs` list comprehension — neither of the described `output_ids` fixes was implemented. Both primary bugs from issue #199 remain present in the code. Additionally, the single guard that was added (`n["id"] in keras_tensors` on the `inputs` line) introduces a silent failure mode where disconnected input nodes are dropped without warning, potentially building a structurally wrong model. The PR description is materially inaccurate relative to the diff, and no tests were added to verify any of the edge cases.

**Suggested Version Bump:** `patch`

### ⚠️ Breaking Changes

- **[INFO]** Behavioral change: disconnected custominput nodes silently excluded from model inputs: **Previous behavior**: If a `custominput` node was somehow absent from `keras_tensors` (e.g., due to disconnection from the graph before BFS), the code would raise a `KeyError` when building the `inputs` list.

**New behavior**: Such nodes are silently skipped, and the model is built with fewer inputs.

**Who is affected**: Any caller that passes a `model_params` dict where a `custominput` node is not reachable during BFS traversal. Previously this crashed; now it silently succeeds with a potentially incorrect model.

**Migration path**: No migration needed, but callers should be aware that the model returned may have fewer inputs than nodes of type `custominput` in the input graph. Consider adding input count validation after `model_generation` returns.
  > **Suggestion:** This is a patch-level behavioral change (crash → silent degradation), but it changes the observable contract of `model_generation`. Consider documenting the new behavior or logging a warning when nodes are skipped.


### PR Description Issues

- **[MEDIUM]** PR description describes changes that were not actually made: The PR description states:

> Fixed by adding two guards to the **output_ids list comprehension**:
> - `n["type"] != "custominput"`: excludes input nodes
> - `n["id"] in keras_tensors`: excludes disconnected unvisited nodes

However, the actual diff shows:
1. **Only one guard was added**, not two.
2. The guard was added to the **`inputs` list comprehension**, not `output_ids`.
3. The `output_ids` line was **not modified at all**.

This is a significant description-to-diff mismatch. It creates confusion for reviewers trying to validate correctness, and it also suggests the PR may have been submitted from the wrong branch or with an incomplete set of changes.
  > **Suggestion:** Update the PR description to accurately reflect the actual change:

> **Actual change**: Added `and n["id"] in keras_tensors` guard to the `inputs` list comprehension to prevent a `KeyError` if a `custominput` node is absent from `keras_tensors`.
>
> **Note**: The fixes to `output_ids` described in this PR (excluding disconnected nodes and custominput nodes from outputs) have **not yet been implemented** and should be addressed in a follow-up or by updating this branch.
- **[LOW]** PR description lacks testing approach / reproduction steps: The PR description links to issue #199 and explains the two problems well, but does not include:
- Steps to reproduce the original bugs
- How the fix was manually validated
- Any mention of automated tests
- Notes on whether existing tests pass

For a bug fix PR that touches model construction logic, verifying behavior both before and after the fix is important for reviewer confidence.
  > **Suggestion:** Add a **Test Plan** section such as:

> ## Test Plan
> - [ ] Manually reproduced KeyError by submitting a model with a disconnected node — confirmed crash before fix, no crash after
> - [ ] Confirmed `custominput` with no outgoing edges is excluded from `outputs`
> - [ ] All existing model generation tests pass (`pytest tests/test_model_generation.py`)
> - [ ] Added regression tests for both scenarios in `tests/test_model_generation.py`

**Comment by ivantha** [comment]
This PR is behind the base branch. Please rebase onto the latest `main` so we can re-run CI and merge.

**Comment by ivantha** [comment]
Thank you for the disconnected-node fix (#199)! This PR has been superseded by #245 which addresses the same issue with an explicit ValidationError raise and broader coverage. We appreciate the careful analysis of the output_ids bug.
